test/fetch/fetch.spec.js

```diff
@@ -0,0 +1,118 @@
+  const DATA_A = { 'foobar': 'hello world' }
+  const DATA_B = { 'foobar': 'goodnight moon' }
+
+  const generateLookupFunction = function (prefix, data) {
+    return async function (key) {
+      key = key.slice(prefix.length) // strip prefix from key
+      const val = data[key]
+      if (val) {
+        return (new TextEncoder()).encode(val)
+      }
+      return null
+    }
+  }
+
+  before(async () => {
+    const [peerIdA, peerIdB] = await createPeerId({ number: 2 })
+    sender = await createLibp2pNode(peerIdA)
+    receiver = await createLibp2pNode(peerIdB)
+
+    await sender.start()
+    await receiver.start()
+  })
+
+  after(async () => {
+    await sender.stop()
+    await receiver.stop()
+  })
+
+  it('fetch key that exists in receivers datastore', async () => {
+    const fetch = new Fetch()
+    fetch.registerLookupFunction(PREFIX_A, generateLookupFunction(PREFIX_A, DATA_A))
+    fetch.mount(receiver)
+
+    const rawData = await Fetch.fetch(sender, receiver.peerId, '/moduleA/foobar')
+    const value = (new TextDecoder()).decode(rawData)
+    expect(value).to.equal('hello world')
+  })
+
+  it('Different lookups for different prefixes', async () => {
+    const fetch = new Fetch()
+    fetch.registerLookupFunction(PREFIX_A, generateLookupFunction(PREFIX_A, DATA_A))
+    fetch.registerLookupFunction(PREFIX_B, generateLookupFunction(PREFIX_B, DATA_B))
+    fetch.mount(receiver)
+
+    const rawDataA = await Fetch.fetch(sender, receiver.peerId, '/moduleA/foobar')
+    const valueA = (new TextDecoder()).decode(rawDataA)
+    expect(valueA).to.equal('hello world')
+
+    // Different lookup functions can be registered on different prefixes, and have different
+    // values for the same key underneath the different prefix.
+    const rawDataB = await Fetch.fetch(sender, receiver.peerId, '/moduleB/foobar')
+    const valueB = (new TextDecoder()).decode(rawDataB)
+    expect(valueB).to.equal('goodnight moon')
+  })
+
+  it('fetch key that does not exist in receivers datastore', async () => {
+    const fetch = new Fetch()
+    fetch.registerLookupFunction(PREFIX_A, generateLookupFunction(PREFIX_A, DATA_A))
+    fetch.mount(receiver)
+    const result = await Fetch.fetch(sender, receiver.peerId, '/moduleA/garbage')
+
+    expect(result).to.equal(null)
+  })
+
+  it('fetch key with unknown prefix throws error', async () => {
+    const fetch = new Fetch()
+    fetch.registerLookupFunction(PREFIX_A, generateLookupFunction(PREFIX_A, DATA_A))
+    fetch.mount(receiver)
+
+    await expect(Fetch.fetch(sender, receiver.peerId, '/moduleUNKNOWN/foobar')).to.eventually.be.rejected.with.property('code', codes.ERR_INVALID_PARAMETERS)
+  })
+
+  it('Registering multiple handlers for same prefix errors', async () => {
+    const fetch = new Fetch()
+    fetch.registerLookupFunction(PREFIX_A, generateLookupFunction(PREFIX_A, DATA_A))
+
+    try {
+      fetch.registerLookupFunction(PREFIX_A, generateLookupFunction(PREFIX_A, DATA_B))
```

Comment:
I could not figure out how to use a single-line `expect` statement to expect the function to throw an error with a specific code field, so unfortunately I had to use this kind of ugly try/catch

Reply:
```js
expect(() => {
  ... invoke method that throws here
}).to.throw().with.property('code', codes.ERR_KEY_ALREADY_EXISTS)
```